format

```diff
@@ -1,27 +1,51 @@
 #! /usr/bin/env nix-shell
-#! nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/6616de389ed55fba6eeba60377fc04732d5a207c.tar.gz -i bash -p findutils nixfmt
+#! nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/6616de389ed55fba6eeba60377fc04732d5a207c.tar.gz -i bash -p git gnugrep gnused findutils nixfmt
 
-CHECK_ARG=
+nixfmt_args=()
+files=()
 
-case $1 in
-    -h)
-        echo "$0 [-c]"
-        ;;
-    -c)
-        CHECK_ARG=-c
-        ;;
-esac
+for arg do
+    case $arg in
+        -h)
+            echo "$0 [-c]"
+            exit
+            ;;
+        -c)
+            nixfmt_args+=("$arg")
+            ;;
+        -*)
+            echo "unrecognised flag: $arg" >&2
+            exit 1
+            ;;
+        *)
+            files+=("$arg")
+            ;;
+    esac
+done
 
 # The excludes are for files touched by open pull requests and we want
 # to avoid merge conflicts.
-find . -name '*.nix' \
-  ! -path ./modules/default.nix \
-  ! -path ./modules/files.nix \
-  ! -path ./modules/home-environment.nix \
-  ! -path ./modules/lib/default.nix \
-  ! -path ./modules/lib/file-type.nix \
-  ! -path ./modules/misc/news.nix \
-  ! -path ./modules/programs/ssh.nix \
-  ! -path ./modules/programs/zsh.nix \
-  ! -path ./tests/default.nix \
-  -exec nixfmt $CHECK_ARG {} +
+excludes=(
+    modules/default.nix
+    modules/files.nix
+    modules/home-environment.nix
+    modules/lib/default.nix
+    modules/lib/file-type.nix
+    modules/misc/news.nix
+    modules/programs/ssh.nix
+    modules/programs/zsh.nix
+    tests/default.nix
+)
+
+exclude_args=()
+for e in "${excludes[@]}"; do
+    exclude_args+=(-e "$e")
+done
+
+git_root=$(git rev-parse --show-toplevel)
+
+git ls-files --full-name -- "${files[@]}" |
+    grep '\.nix$' |
+    grep -v "${exclude_args[@]}" |
+    sed "s|^|$git_root/|" |
+    xargs -d '\n' nixfmt "${nixfmt_args[@]}"
```

Comment:
Probably best to check that `sed` command. I don't think that's the right syntax.

Either way, I think `find` was the right approach anyways, as `git ls-files` will only work on tracked files. Now, that may actually be desirable behavior, but I doubt it. If it is, there are options to `git ls-files` for exclusions, null-termination.

It'd make much more sense afaict to keep the default behavior of formatting all files in the repository if no file names are provided, but to run the formatter absolutely if given any file names.

That being said, the `find` function could well leverage `-execdir` instead of `exec`, and ensuring `-type f`. It might actually be faster though to leverage the ability of `nixfmt` to work on multiple files with `find $GITROOT -mindepth 1 -name '.git' -prune -o -type f $FINDEXPR[@] -print0 | xargs -0 nixfmt ${nixfmt_args[@]}`. Where $FINDEXPR is a statement like `! -path $FILEPATH`.

GNU find is quite a beast...

Reply:
> Probably best to check that `sed` command. I don't think that's the right syntax.

> :sparkles: Always read the manual! You never know what you might learn! :sparkles:

> as `git ls-files` will only work on tracked files

We can add `-co`, that would list untracked files as well.

Sadly `git ls-files --exclude` only filters untracked files, so we have to do the filtering ourselves. (We can't use [pathspecs](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec) for this either: they're understood as a disjunction, so we can't say "list the files under those directories AND that match `*.nix`").

> null-termination

I'm normally in favour of that but that's obviously never going to be an issue with home-manager and I can't be bothered.

> `find ... -print0 | xargs -0 nixfmt ...`

That's equivalent to `-exec nixfmt {} +`, which is what we had previously.

Reply:
> can't be bothered.

Did it anyway.

Reply:
> > Probably best to check that `sed` command. I don't think that's the right syntax.
> 
> > sparkles Always read the manual! You never know what you might learn! sparkles

Guess that's what I get for not reading every possible manpage out there and expecting GNU's `sed` manpage to actually detail the command, not just implementation-specifics...

Reply:
Yeah GNU man pages suck. This is documented in the POSIX man page `sed(1p)` and in the actual GNU sed manual.